spec/rubocop/cop/rspec/leaked_const_spec.rb

```diff
@@ -0,0 +1,100 @@
+RSpec.describe RuboCop::Cop::RSpec::LeakedConst do
+  subject(:cop) { described_class.new }
+
+  it 'finds a new named class defined in-line' do
+    expect_violation(<<-RUBY)
+      class Foo
+      ^^^^^^^^^ Opening a class to define methods can pollute your tests. Instead, try using `stub_const` with an anonymized class.
+      end
+    RUBY
+  end
+
+  it 'finds a new named class defined inside of a let block' do
+    expect_violation(<<-RUBY)
+      let(:fake_class) do
+        class Foo
+        ^^^^^^^^^ Opening a class to define methods can pollute your tests. Instead, try using `stub_const` with an anonymized class.
+        end
+      end
+    RUBY
+  end
+
+  it 'finds a new named class assigned to a variable' do
+    expect_violation(<<-RUBY)
+      fake_class = class Foo
+                   ^^^^^^^^^ Opening a class to define methods can pollute your tests. Instead, try using `stub_const` with an anonymized class.
+      end
+    RUBY
+  end
+
+  it 'finds a new unnamed class inside of a let block' do
+    expect_no_violations(<<-RUBY)
+      let(:fake_class) { Class.new }
+    RUBY
+  end
+
+  it 'finds a new unnamed class assigned to a variable' do
+    expect_no_violations(<<-RUBY)
+      fake_class = Class.new
+    RUBY
+  end
+
+  it 'finds a new class already marked with stub_const' do
+    expect_no_violations(<<-RUBY)
+      before do
+        stub_const('Foo', Class.new)
+        class Foo
+        end
+      end
+    RUBY
+  end
+
+  it 'finds a new class in a module that is already marked with stub_const' do
+    expect_no_violations(<<-RUBY)
+      before do
+        stub_const('Namespace', Module.new)
+        module Namespace
+          class Foo
+          end
+        end
+      end
+    RUBY
+  end
+
+  it 'finds two new classes in a module already marked with stub_const' do
+    expect_no_violations(<<-RUBY)
+      before do
+        stub_const('Namespace::SecondNamespace', Module.new)
+        module Namespace
+          module SecondNamespace
+            class Foo
+            end
+
+            class Bar
+            end
+          end
+        end
+      end
+    RUBY
+  end
+
+  it 'finds a new class marked for stub_const after opening the class' do
+    expect_violation(<<-RUBY)
+      before do
+        class Foo
+        ^^^^^^^^^ Opening a class to define methods can pollute your tests. Instead, try using `stub_const` with an anonymized class.
+        end
+        stub_const('Foo', Class.new)
+      end
+    RUBY
+  end
+
+  it 'finds a basic test set-up using modules to open the describe block' do
+    expect_no_violations(<<-RUBY)
+      module TestedNamespace
```

Comment:
Perhaps this should be a different cop, but we should be flagging this kind of test too, imo. This is probably not a good idea because the interior of the module (or class) is open/mutable and you are now scoped in such a way that you can reference private implementation details of the module that should not be tested directly.

Regardless of whether we should do that here, we should definitely flag module declarations inside of the describe like this:

```ruby
RSpec.describe MyClass do
  module TestModule
    def self.foo
      'bar'
    end
  end

  it 'somehow uses the module' do
    expect(TestModule.foo).to eql('bar')
  end
end
```

as it has the same global setup problem that a class declaration does.